### ISSUE TITLE
URGENT: Possible vulnerability to injection attacks

### ISSUE THREAD

**Issue by jimcullenaus** [naturalcrit/homebrewery#546]
Problem: inserting arbitrary JavaScript into documents is possible, and it will execute. It also seems to cause the entire rest of the page to not render. The person with edit permission to the document cannot then *remove* the JS, because the edit field is also blanked out.

Solution: disable `<script>` tags. Do not allow execution of any JavaScript entered by the users.

For an example, try creating a new blank document, entering some dummy text, and then pasting this at the bottom:

`<script>
document.title = "The title";
</script>`
Or, see [this document](http://homebrewery.naturalcrit.com/share/SkHdIxW00W) (and its source, which is available).

Note that the title in your web browser changes.

**Comment by dekelpilli** [comment]
Was just about to report this. In addition to what is specified here, it seems to mess with the user page of whoever owns the page. The following is a screenshot of my user page after I added a script to one of my brews -- https://i.imgur.com/0cdimPi.jpg -- these are all images I've used in my brews. Below the normal page, it also has some plaintext version of some of my brews (\ns, \ts, and all) -- https://i.imgur.com/4q61IaH.png

Also, the screenshots might look yellow, that part is on my end, not homebrewery's.

**Comment by goopscoop** [comment]
I solved this in a different project with the `sanitize-html` node module. Super easy to setup. Want me to take care of this for you?

**Comment by kkragenbrink** [comment]
Of course, the problem is there's also legitimate reasons to add Javascript, such as zoomable images, or the like.  This is not as easy to fix as simply disabling script tags, but I'll look into it.

**Comment by jimcullenaus** [comment]
I cannot see any legitimate reason to use JavaScript in a homebrewery document. Homebrewery is for building up pages that look high-quality like the D&D published books. There's no room for dynamic content in that. All it does is increase the attack surface.

**Comment by Sansd20** [comment]
I concur with @jimcullenaus. While the homebrewery documents do have the ability to link, the over all use case is primarily to create static documents.

If you are needing Javascript for your use case, you might want to fork this project (if that is possible) and come up with your own solution.

**Comment by AlexeySachkov** [comment]
Should be fixed by #811

**Comment by calculuschild** [comment]
Closing because this appears to be solved.

**Comment by 5e-Cleric** [comment]
Reopening because a closing script tag still breaks the website, see #3434

**Comment by dbolack-ab** [comment]
Is this resolved by the DOMPurify addition?

**Comment by 5e-Cleric** [comment]
> Is this resolved by the DOMPurify addition?

![image](https://github.com/naturalcrit/homebrewery/assets/91662974/9e29f837-87ae-44f2-8686-6d0dfd7686de)

Nope

**Comment by ericscheid** [comment]
DOMpurify etc are not going to fix this.

This is what is happening..

1. the brew source is combined with the style source, metadata, and other bits into a javascript object
2. this object then is put through `JSON.stringify()`
3. and the resulting string is then included the http-response payload as a part of the html
4. it appears like this: `<script>start_app({"version":"3.12.0","url":"/edit/5M-3mB5QPvE2","brew":{"text":"brew text </script>","pageCount":"1","renderer":"V3", [etc] })</script>`
5. this is then delivered to the browser as a http-response
6. the browser then parses the response into html elements
7. and finally it renders the html, and runs the javascript `start_app(...)`

The subtle trickiness here though is that in step 6, when the browser is parsing html, it sees `<script>` and says "ah, this is a script element .. which runs until a closing `</script>` tag appears". The problem is that the parser sees the string `</script>` prematurely, and says "ah, end of the `<script>` element. Any text it finds after point is outside of the script element. Then, in step 7, when the browser goes to run the javascript it sees `<script>[the script]</script>`, where [the script is all this text: `start_app({"version":"3.12.0","url":"/edit/5M-3mB5QPvE2","brew":{"text":"brew text `. That is not valid javascript, and thus we get the exception of `Uncaught SyntaxError: Invalid or unexpected token`. The rest of the intended script has been cut off by the parser.

So, in [`template.js`](https://github.com/naturalcrit/homebrewery/blob/master/client/template.js) we need to defang the string we pass in to `start_app`. 

Change line 26 to
`<script>start_app(${JSON.stringify(props).replace(/</g,"&lt;"})</script>`

**Comment by dbolack-ab** [comment]
I can no longer replicate this issue.

**Comment by ericscheid** [comment]
> I can no longer replicate this issue.

I can.

- Inserting `</script>` into the home page content doesn't trigger it
- Inserting `</script>` into the `/new` page content doesn't trigger it
- Reloading the `/new` editor doesn't trigger it
  - it's dynamically loading the brew content from the `localStorage` property
  - not receiving it as a string within the html source
- Saving the `/new` containing `</script>` **_does_** trigger it
  - because saving `/new` redirects to loading the `/edit/...` url
- Reloading `/edit/...` triggers it
- Loading `/preview/...` triggers it
- **hmmm .. loading `/user/erics` triggers it ?!?**
  - ah .. saving from `/new` set the brew.title property to "</script>", and that kills 
  - this also implies that `brew.description` and other `brew.meta` properties could also kill. Including `brew.authors`.

The reason for the fail is the same as described in https://github.com/naturalcrit/homebrewery/issues/546#issuecomment-2149771521 above. 

<img width="663" height="36" alt="Image" src="https://github.com/user-attachments/assets/f1da1da6-aa55-475e-8c1b-8536862783a0" />

<img width="968" height="36" alt="Image" src="https://github.com/user-attachments/assets/83a0c19b-9523-4e7f-a47d-ce65948186da" />

**Comment by 5e-Cleric** [comment]
Is this fixed by the allow-scripts PR? @ReinaSweet @calculuschild

**Comment by 5e-Cleric** [comment]
Sorry, no, can confirm this still breaks the site with a simple `</script>` in edit

**Comment by ericscheid** [comment]
> It also seems to cause the entire rest of the page to not render

Don't know if I've mentioned this yet .. this isn't a script permissions issue, it's a html parsing issue.

The browser receives a payload of "text/html" which it then attempts to parse. When it sees "<script>" in the source it starts looking for "</script>" and treats everything in between as the contents of a `<script>` DOM element. Which gets handed off to the javascript parser, which then borks on a chunk of text that isn't valid javascript.

The browser sends 

> `<html>...<script>window.__INITIAL_PROPS__ = {"version":"3.23.0", ... ,"invitedAuthors":[],"googleId":null,"lang":"en","text":"brew text containing the string </script>", ... }}</script>`, 

.. and from that extracts this text...

> `<script>window.__INITIAL_PROPS__ = {"version":"3.23.0", ... ,"invitedAuthors":[],"googleId":null,"lang":"en","text":"brew text containing the string </script>`

.. which is _invalid_ javascript, and so the console reports an error and javascript processing for the page is halted.

> Uncaught SyntaxError: Invalid or unexpected token at ...

----

The solution to this issue is to go to `app.js` and stop sending raw "</script>" strings in the JSON.stringify(props).

https://github.com/ericscheid/homebrewery/blob/7f47fe0657ccc7fb38aa98adb470d90b93e77dea/server/app.js#L605-L608

```js
html = html.replace(
	'<head>',
	()=>{ return `<head>\n<script id="props" >window.__INITIAL_PROPS__ = ${JSON.stringify(props)}</script>\n${ogMetaTags}`; }
);
```

The easiest way is to replace the `<` character with it's json unicode reference. Like this:

```js
const safeProps = JSON.stringify(props).replace(/<\/script/ig, '\\u003c/script');
```

BTW, doing this to all of `props` will also defang any `</script` strings in brew.title, brew.description, etc..

**Comment by 5e-Cleric** [comment]
If you open that PR, i can review that straight away.

**Comment by ericscheid** [comment]
> If you open that PR, i can review that straight away.

It's late here .. probably best I don't open a PR. Sleep first =)
